README.md
## Laravel Alpine Animations

[![Packagist License](https://poser.pugx.org/Jubeki/laravel-alpine-animations/license.png)](http://choosealicense.com/licenses/mit/)
[![Latest Stable Version](https://poser.pugx.org/Jubeki/laravel-alpine-animations/version.png)](https://packagist.org/packages/jubeki/laravel-alpine-animations)
<!-- ![Tests](https://github.com/Jubeki/laravel-alpine-animations/workflows/Tests/badge.svg) -->

## Installation

Require this package with composer.

```shell
composer require jubeki/laravel-alpine-animations
```

The service provider will be automatically registered using [package discovery](https://laravel.com/docs/5.8/packages#package-discovery).

If you don't use auto-discovery you should add the service provider to the providers array in `config/app.php`.

```php
// existing providers...
Jubeki\AlpineAnimations\AlpineAnimationsServiceProvider::class,
```

The standard styles are written with Tailwind CSS. If you want to make sure that all the styles are added to your resulting CSS File you should add the following path to your `tailwind.config.js`:
```js
module.exports = {
    content: [
        "./vendor/jubeki/laravel-alpine-animations/resources/**/*.blade.php"
    ],
    theme: {
        extend: {},
    },
    plugins: [],
}
```

## Usage

You can now start using all available components using the standard blade components syntax.

If a user doesn't have JavaScript enabled, then the default option will be shown.

## Components

### Typing

![](examples/typing-with-cursor.gif)

Minimal Example:
```blade
<x-alpine-animations::typing
    :values="['First Line', 'Second Line', 'Third Line']"
/>
```

Maximal Example (with Default values):
```blade
<x-alpine-animations::typing
    :values="['First Line', 'Second Line', 'Third Line']"
    start-at="0"
    init-delay="700"
    deleting-delay="70"
    typing-delay="100"
    start-deleting-delay="1200"
    start-typing-delay="300"
    toggle-cursor-delay="375"
    cursor="border-r-4"
    cursor-hidden="border-transparent"
    cursor-shown="border-black"
/>
```

![](examples/typing-without-cursor.gif)

You can also disable the cursor altogether which leaves only the typing and deleting of characters:
```blade
<x-alpine-animations::typing
    :values="['First Line', 'Second Line', 'Third Line']"
    :cursor="false"
/>
```

### Magnetic

<!-- ![](examples/magnetic.gif) -->

Minimal Example:
```blade
<x-alpine-animations::magnetic
    inner-class="inline-flex rounded-full bg-indigio-600 items-center justify-center w-14 h-14"
>
    &rightarrow;
</x-alpine-animations::magnetic>
```

Maximal Example (with Default values, except `inner-class`):
```blade
<x-alpine-animations::magnetic
    inner-class="inline-flex rounded-full bg-indigio-600 items-center justify-center w-14 h-14"
    as="button"
    translate-x-modifier="0.7"
    translate-y-modifier="0.7"
>
    &rightarrow;
</x-alpine-animations::magnetic>
```

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](./.github/CONTRIBUTING.md) for details.

## Credits

- [Julius Kiekbusch](https://github.com/Jubeki)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.

## Unreleased

### Added
### Changed 
### Removed 

## 0.2.0

### Added

- Added Magnetic Button Animation

## 0.1.1

### Added

- Added ability to disable cursor for typing animation

## 0.1.0

### Added

- Added Package Structure for initial release
- Added Typing Animation


resources/views/components/magnetic.blade.php
@props([
    'as' => 'button',
    'innerClass' => '',
    'translateXMultiplier' => 0.7,
    'translateYMultiplier' => 0.7,
])

<{{ $as }}
    x-data="{
        translateXMultiplier: {{ $translateXMultiplier }},
        translateYMultiplier: {{ $translateYMultiplier }},

        move(e) {
            const position = this.$root.getBoundingClientRect()

            const translateX = (e.x - position.left - position.width / 2) * this.translateXMultiplier
            const translateY = (e.y - position.top - position.height / 2) * this.translateYMultiplier

            this.$root.children[0].style.transform = `translate(${translateX}px, ${translateY}px)`
        },

        reset(e) {
            this.$root.children[0].style.transform = 'translate(0px, 0px)'
        },
    }"
    @@mousemove="move($event)"
    @@mouseout="reset($event)"
    {{ $attributes->class(['relative cursor-pointer'])}}
>
    <span class="{{ $innerClass }} relative m-10 transition-transform ease-out duration-1000">
        {{ $slot }}
    </span>
</{{ $as }}>


resources/views/components/typing.blade.php
@props([
    'values',
    'startAt' => 0,
    'initDelay' => 700,
    'deletingDelay' => 70,
    'typingDelay' => 100,
    'startDeletingDelay' => 1200,
    'startTypingDelay' => 300,
    'toggleCursorDelay' => 375,
    'cursor' => 'border-r-4',
    'cursorHidden' => 'border-transparent',
    'cursorShown' => 'border-black',
])

<span
    @if($cursor !== false)
        {{ $attributes->class($cursor.' '.$cursorHidden) }}
        :class="{
            '{{ $cursorHidden }}': !showCursor,
            '{{ $cursorShown }}': showCursor,
        }"
    @else
        {{ $attributes }}
    @endif
    x-text="text"
    x-data="{
        showCursor: false,
        values: @js($values),
        current: @js($startAt),
        text: @js($values[$startAt]),
        initDelay: {{ $initDelay }},
        deletingDelay: {{ $deletingDelay }},
        typingDelay: {{ $typingDelay }},
        startDeletingDelay: {{ $startDeletingDelay }},
        startTypingDelay: {{ $startTypingDelay }},
        @if($cursor !== false)
        toggleCursorDelay: {{ $toggleCursorDelay }},
        toggleCursor() {
            this.showCursor = !this.showCursor
        },
        @endif
        type() {
            this.text = this.values[this.current].substring(0, this.text.length + 1)

            if(this.text === this.values[this.current]) {
                setTimeout(() => this.delete(), this.startDeletingDelay)
            } else {
                setTimeout(() => this.type(), this.typingDelay)
            }
        },
        delete() {
            this.text = this.text.slice(0, -1)

            if(this.text === '') {
                this.current = (this.current + 1) % this.values.length
                this.current_curr = 0

                setTimeout(() => this.type(), this.startTypingDelay)
            } else {
                setTimeout(() => this.delete(), this.deletingDelay)
            }
        },
        init() {
            setTimeout(() => this.delete(), this.initDelay)
            @if($cursor !== false)
            setInterval(() => this.toggleCursor(), this.toggleCursorDelay)
            @endif
        },
    }"
>{{ $values[$startAt] }}</span>
